crates/consensus/src/round.rs

```diff
@@ -147,384 +112,232 @@ impl RoundStateMachine {
-
-        unique_limit_per_pool
+        input: Vec<OrderWithStorageData<T>>
+    ) -> Vec<OrderWithStorageData<T>> {
+        input
             .into_iter()
-            .flat_map(|(pool_id, orders)| {
-                orders.into_iter().map(move |order| OrderWithStorageData {
-                    pool_id,
-                    order,
-                    ..Default::default()
-                })
+            .fold(HashMap::new(), |mut acc, order| {
+                *acc.entry(order).or_insert(0) += 1;
+                acc
             })
+            .into_iter()
+            .filter(|(_, count)| self.has_quorum(*count))
+            .map(|(order, _)| order)
             .collect()
     }
 
-    fn merge_searcher_orders(
-        &self,
-        left_searcher: Vec<OrderWithStorageData<TopOfBlockOrder>>
-    ) -> Vec<OrderWithStorageData<TopOfBlockOrder>> {
-        let mut top_searcher_per_pool: HashMap<_, OrderWithStorageData<TopOfBlockOrder>> =
-            HashMap::new();
-        for order_with_data in left_searcher {
-            let pool_id = order_with_data.pool_id;
-            top_searcher_per_pool
-                .entry(pool_id)
-                .and_modify(|existing_order| {
-                    if order_with_data.tob_reward > existing_order.tob_reward {
-                        *existing_order = order_with_data.clone();
-                    }
-                })
-                .or_insert(order_with_data);
-        }
-        top_searcher_per_pool.into_values().collect()
-    }
-
     fn force_transition(&mut self, mut new_state: ConsensusState) {
-        let from_state = self.current_state.clone();
         let signer = self.signer.clone();
+        let metrics = self.metrics.clone();
+        let pre_proposal_height = self.current_state.block_height();
+        let pre_proposals: Vec<PreProposal> =
+            self.current_state.pre_proposals().iter().cloned().collect();
 
         self.transition_future = Some(Box::pin(async move {
-            if let (
-                ConsensusState::BidAggregation(BidAggregation {
-                    block_height: pre_proposal_height,
-                    pre_proposals
-                }),
-                ConsensusState::Finalization(ref mut finalization)
-            ) = (from_state, &mut new_state)
-            {
-                let pre_proposals: Vec<PreProposal> = pre_proposals.iter().cloned().collect();
-                // TODO: maybe spawn a new thread here, since proposal building might be comp
-                // expensive
-                let solutions = build_proposal(pre_proposals.clone()).await.unwrap();
-                let proposal = signer.sign_proposal(pre_proposal_height, pre_proposals, solutions);
-
-                finalization.proposal = Some(proposal);
-
-                // TODO: submit the proposal/bundle to the chain
-                // Ethereum.submit(proposal).await
+            if let ConsensusState::Finalization(ref mut finalization) = &mut new_state {
```

Comment:
idt this `ref mut` is necessary - just a semantics thing

```suggestion
            if let ConsensusState::Finalization(finalization) = &mut new_state {
```